README.md
# OTS SDK [![Build Status](https://secure.travis-ci.org/fengmk2/ots.png)](http://travis-ci.org/fengmk2/ots) [![Coverage Status](https://coveralls.io/repos/fengmk2/ots/badge.png)](https://coveralls.io/r/fengmk2/ots)

[![NPM](https://nodei.co/npm/ots.png?downloads=true&stars=true)](https://nodei.co/npm/ots)

Aliyun [OTS](http://ots.aliyun.com/)(Open Table Service) SDK for [Node.js](http://nodejs.org).

Using `protobuf` protocol API on `ots@0.4.0+`.

**WANNING** If you are still using `XML` protocol API, please use `ots<=0.3.0`.

## Support API

* TableGroup
    * CreateTableGroup
    * DeleteTableGroup
    * ListTableGroup
* Table
    * CreateTable
    * DeleteTable
    * GetTableMeta
    * ListTable
* Transaction
    * StartTransaction
    * CommitTransaction
    * AbortTransaction
* DataRow
    * PutRow
    * MultiPutRow (Max 100 rows)
    * DeleteRow
    * MultiDeleteRow (Max 100 rows)
    * BatchModifyRow (working)
    * GetRow
    * MultiGetRow (Max 10 rows)
    *
    * GetRowsByRange
    * ~~GetRowsByOffset~~ (removed)

## Install

```bash
$ npm install ots

# If you are still using XML protocol API, please use this:
$ npm install ots@0.3.0
```

## Usage

```js
var ots = require('ots');
var client = ots.createClient({
  accessID: 'your accessID',
  accessKey: 'your accessKey'
});

// create a table
client.createTable({
  TableName: 'testdemo',
  PrimaryKey: [
    {'Name': 'uid', 'Type': 'STRING'},
  ],
  PagingKeyLen: 0,
  View: [
    {
      'Name': 'view1',
      'PrimaryKey' : [
        {'Name':'uid', 'Type':'STRING'},
        {'Name':'flag', 'Type':'STRING'},
        {'Name':'docid', 'Type':'STRING'},
      ],
      'Column' : [
        {'Name':'updatetime', 'Type':'STRING'},
        {'Name':'createtime', 'Type':'STRING'},
      ],
     'PagingKeyLen': 2
    }
  ]
}, function(err, result) {
  console.log(err, result);
});

// insert a row
client.putRow('user',
  { Name: 'uid', Value: 'mk2' },
  [
    { Name: 'firstname', Value: 'yuan' },
    { Name: 'lastname', Value: 'feng\' aerdeng' },
    { Name: 'nickname', Value: '苏千' },
    { Name: 'age', Value: 28 },
    { Name: 'price', Value: 110.5 },
    { Name: 'enable', Value: true },
    { Name: 'man', Value: true },
    { Name: 'female', Value: false },
    { Name: 'createtime', Value: new Date().toJSON() },
  ], function(err, result) {
  console.log(err, result);
});

// get a row
client.getRow('user', { Name: 'uid', Value: 'mk2' }, function (err, row) {
  console.log(err, row);
});
```

More examples, please see [test/client.test.js](https://github.com/fengmk2/ots/blob/master/test/client.test.js).

## Links

* [OTS RESTful API Documents](http://ots.aliyun.com/ots_sdk/OTS_RESTful_API_2012_03_22.pdf)
* [OTS Data Model](http://ots.aliyun.com/ots_sdk/OTS_Data%20Model_2012_03_22.pdf)
* [OTS Dashboard](http://ots.aliyun.com/dashboard)
* [Python SDK](http://ots.aliyun.com/ots_sdk/ots_python_sdk_2012_03_22.zip)
* [.Net SDK](http://ots.aliyun.com/ots_sdk/Aliyun_SDK_dotNET_1_0_4458.zip)
* [Java SDK](http://storage.aliyun.com/oss/aliyun_portal_storage/oss_api/OSS_OTS_Java_SDK.zip)
* [OTS Developer Guide](http://ots.aliyun.com/guide/index)

## Authors

```bash
$ git summary

 project  : ots
 repo age : 1 year, 7 months
 active   : 22 days
 commits  : 63
 files    : 23
 authors  :
    58  fengmk2                 92.1%
     3  tangyao                 4.8%
     2  coolme200               3.2%
```

## License

(The MIT License)

Copyright (c) 2012 - 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.5.2 / 2014-02-18 
==================

  * fix test cases
  * fix OTS endpoint to be the latest one: http://ots.aliyuncs.com (@edwardliwang)
  * add npm image

0.5.1 / 2013-11-05 
==================

  * fix requestNormal undefined bug

0.5.0 / 2013-11-05 
==================

  * support importdata2 api fixed #10

0.4.9 / 2013-11-05 
==================

  * convertToColumnValue() support all data type

0.4.8 / 2013-11-05 
==================

  * only left debug url when error happend
  * add more cases

0.4.7 / 2013-11-05 
==================

  * fix ots bigint

0.4.6 / 2013-10-29 
==================

  * support vip manager
  * improve benchmark codes
  * add benchmark script

0.4.5 / 2013-10-15 
==================

  * handle Malformed message fixed #9

0.4.4 / 2013-10-15 
==================

  * use protobuf-ali, make int64 to number when it small
  * add more get rows by range cases

0.4.3 / 2013-10-15 
==================

  * fix parseRow() bug
  * update doc

0.4.2 / 2013-09-25 
==================

  * batchModifyRow() done, remove GetRowsByOffset(). fixed #8
  * GetRowsByRange() done

0.4.1 / 2013-09-25 
==================

  * multiDeleteRow()
  * MultiPutRow()

0.4.0 / 2013-09-24 
==================

  * Using `protobuf` protocol API. MultiGetRow(), deleteRow()
  * putRow() and getRow()
  * transation, table, tablegroup
  * fix test cases
  * fix tests

0.3.0 / 2013-08-20 
==================

  * use xml2json instead xml2json-edp

0.2.0 / 2013-07-01 
==================

  * add cache dns support

0.1.9 / 2013-04-14 
==================

  * fixed getRow typeerror

0.1.8 / 2013-04-14 
==================

  * fixed TypeError on `Client.prototype.parseRows`

0.1.7 / 2013-04-09 
==================

  * update error info

0.1.6 / 2013-04-08 
==================

  * fixed require bug

0.1.5 / 2013-04-08 
==================

  * use xml2json-edp 0.2.5

0.1.4 / 2013-04-08 
==================

  * Merge pull request #2 from coolme200/patch-1
  * upgrade version 0.1.4
  * Merge pull request #1 from coolme200/issue-ots-error
  * revers xml2json version 0.2.4
  * add sanitize for xml value
  * add travis.xml


test/xml2json.test.js
// /*!
//  * ots - test/xml2json.test.js
//  * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
//  * MIT Licensed
//  */

// "use strict";

// /**
//  * Module dependencies.
//  */

// var xml2json = require('xml2json');
// var should = require('should');
// var fs = require('fs');
// var xml2jsonOptions = require('../').xml2jsonOptions;

// var fixturesDir = __dirname + '/fixtures';

// describe('xml2json', function () {
//   it('should return string space', function () {
//     // https://github.com/buglabs/node-xml2json/issues/23
//     var xml = fs.readFileSync(fixturesDir + '/ots-result.xml', 'utf-8');
//     var json = xml2json.toJson(xml, xml2jsonOptions);
//     // console.log('%j', json);

//     json.should.have.property('doc');
//     json.doc.Column.should.length(5);
//     json.doc.Column[0].should.eql({Name: 'shit', Value: {type: 'STRING', $t: '  abc\nasdf\na '}});
//     json.doc.Column[1].should.eql({Name: 'foo', Value: {type: 'STRING'}});
//     json.doc.Column[2].should.eql({Name: 'foo2', Value: {type: 'STRING'}});
//     json.doc.Column[3].should.eql({Name: 'bar', Value: {type: 'STRING', $t: ' '}});
//     json.doc.Column[4].should.eql({PK: 'true', Name: 'uid', Value: {type: 'STRING', $t: 'god'}});
//   });

//   it('should success quot', function (done) {
//     var xml = fs.readFileSync(fixturesDir + '/ots-result2.xml', 'utf-8');
//     var json = xml2json.toJson(xml, xml2jsonOptions);
//     JSON.stringify(json).should.include('\\"isAward\\":false');
//     json.should.have.keys('GetRowResult');
//     json.GetRowResult.should.have.keys('Table', 'RequestID', 'HostID');
//     json.GetRowResult.Table.Row.should.have.keys('Column');
//     json.GetRowResult.Table.Row.Column.should.eql([
//       { Name: 'user', Value: { type: 'STRING', '$t': '1' } },
//       { Name: 'member_car_info', Value: { type: 'STRING', '$t': '{"isAward":false}' } } 
//     ]);
//     done();
//   });
// });




benchmark.js
var Benchmark = require('benchmark');
var xml2json = require('xml2json');
var fs = require('fs');
var suite = new Benchmark.Suite();

var xml = '\
<doc>\
  <Column>\
    <Name>shit</Name>\
    <Value type="STRING">  abc\
asdf\
a </Value>\
  </Column>\
  <Column><Name>foo</Name><Value type="STRING"></Value></Column>\
  <Column><Name>foo2</Name><Value type="STRING" /></Column>\
  <Column><Name>bar</Name><Value type="STRING"> </Value></Column>\
  <Column PK="true"><Name>uid</Name><Value type="STRING">god</Value></Column>\
</doc>\
';

suite
.add('toJson() with {object: true, space: true}', function () {
  var json = xml2json.toJson(xml, {object: true, space: true});
})
.add('toJson() with {object: true, space: false}', function () {
  var json = xml2json.toJson(xml, {object: true, space: false});
})
.add('toJson() with {}', function () {
  var json = xml2json.toJson(xml);
})
// add listeners
.on('cycle', function (event, bench) {
  console.log(String(bench));
})
.on('complete', function () {
  console.log('Fastest is ' + this.filter('fastest').pluck('name'));
})
.run();



benchmark/parse.js
/*!
 * ots - benchmark/parse.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var Benchmark = require('benchmark');
var xml2json = require('xml2json');
var xml2jsonEDP = require('xml2json-edp');
var fs = require('fs');
var path = require('path');
var Schema = require('protobuf-ali').Schema;

var schema = new Schema(fs.readFileSync(path.join(path.dirname(__dirname), 'ots_protocol.desc')));
var Row = schema['com.aliyun.cloudservice.ots.Row'];
var suite = new Benchmark.Suite();

var pbRowBuffer = Row.serialize({
  primaryKeys: [{
    name: 'uid',
    value: {
      type: 'STRING',
      valueS: 'god'
    }
  }],
  columns: [{
    name: 'shit',
    value: {
      type: 'STRING',
      valueS: '  abc\
asdf\
a '
    }
  }, {
    name: 'foo',
    value: {
      type: 'STRING',
      valueS: ''
    }
  }, {
    name: 'foo2',
    value: {
      type: 'STRING',
      valueS: ''
    }
  }, {
    name: 'bar',
    value: {
      type: 'STRING',
      valueS: ' '
    }
  }]
});

var JSONString = JSON.stringify(Row.parse(pbRowBuffer));
// console.log('%j, json: %j', Row.parse(pbRowBuffer), JSON.parse(JSONString));

var xml = '\
<doc>\
  <Column>\
    <Name>shit</Name>\
    <Value type="STRING">  abc\
asdf\
a </Value>\
  </Column>\
  <Column><Name>foo</Name><Value type="STRING"></Value></Column>\
  <Column><Name>foo2</Name><Value type="STRING" /></Column>\
  <Column><Name>bar</Name><Value type="STRING"> </Value></Column>\
  <Column PK="true"><Name>uid</Name><Value type="STRING">god</Value></Column>\
</doc>';

suite
.add('xml2json.toJson() with {object: true, space: true}', function () {
  var json = xml2json.toJson(xml, {object: true, space: true});
})
.add('xml2json.toJson() with {object: true, space: false}', function () {
  var json = xml2json.toJson(xml, {object: true, space: false});
})
.add('xml2json.toJson() with {}', function () {
  var json = xml2json.toJson(xml);
})
.add('xml2jsonEDP.toJson() with {object: true, space: true}', function () {
  var json = xml2jsonEDP.toJson(xml, {object: true, space: true});
})
.add('xml2jsonEDP.toJson() with {object: true, space: false}', function () {
  var json = xml2jsonEDP.toJson(xml, {object: true, space: false});
})
.add('xml2jsonEDP.toJson() with {}', function () {
  var json = xml2jsonEDP.toJson(xml);
})
.add('protobuf.parse()', function () {
  Row.parse(pbRowBuffer);
})
.add('JSON.parse()', function () {
  JSON.parse(JSONString);
})

.on('cycle', function (event) {
  console.log(String(event.target));
})
.on('complete', function () {
  console.log('Fastest is ' + this.filter('fastest').pluck('name'));
})
.run({async: true, delay: 5});



benchmark/xml2json_benchmark.js
var Benchmark = require('benchmark');
var xml2json = require('xml2json');
var xml2jsonEDP = require('xml2json-edp');
var fs = require('fs');
var path = require('path');
var Schema = require('protobuf').Schema;

var schema = new Schema(fs.readFileSync(path.join(path.dirname(__dirname), 'ots_protocol.desc')));
var Row = schema['com.aliyun.cloudservice.ots.Row'];
var suite = new Benchmark.Suite();

var pbRowBuffer = Row.serialize({
  primaryKeys: [{
    name: 'uid',
    value: {
      type: 'STRING',
      valueS: 'god'
    }
  }],
  columns: [{
    name: 'shit',
    value: {
      type: 'STRING',
      valueS: '  abc\
asdf\
a '
    }
  }, {
    name: 'foo',
    value: {
      type: 'STRING',
      valueS: ''
    }
  }, {
    name: 'foo2',
    value: {
      type: 'STRING',
      valueS: ''
    }
  }, {
    name: 'bar',
    value: {
      type: 'STRING',
      valueS: ' '
    }
  }]
});

var JSONString = JSON.stringify(Row.parse(pbRowBuffer));
// console.log('%j, json: %j', Row.parse(pbRowBuffer), JSON.parse(JSONString));

var xml = '\
<doc>\
  <Column>\
    <Name>shit</Name>\
    <Value type="STRING">  abc\
asdf\
a </Value>\
  </Column>\
  <Column><Name>foo</Name><Value type="STRING"></Value></Column>\
  <Column><Name>foo2</Name><Value type="STRING" /></Column>\
  <Column><Name>bar</Name><Value type="STRING"> </Value></Column>\
  <Column PK="true"><Name>uid</Name><Value type="STRING">god</Value></Column>\
</doc>';

suite
.add('xml2json.toJson() with {object: true, space: true}', function () {
  var json = xml2json.toJson(xml, {object: true, space: true});
})
.add('xml2json.toJson() with {object: true, space: false}', function () {
  var json = xml2json.toJson(xml, {object: true, space: false});
})
.add('xml2json.toJson() with {}', function () {
  var json = xml2json.toJson(xml);
})
.add('xml2jsonEDP.toJson() with {object: true, space: true}', function () {
  var json = xml2jsonEDP.toJson(xml, {object: true, space: true});
})
.add('xml2jsonEDP.toJson() with {object: true, space: false}', function () {
  var json = xml2jsonEDP.toJson(xml, {object: true, space: false});
})
.add('xml2jsonEDP.toJson() with {}', function () {
  var json = xml2jsonEDP.toJson(xml);
})
.add('protobuf.parse()', function () {
  Row.parse(pbRowBuffer);
})
.add('JSON.parse()', function () {
  JSON.parse(JSONString);
})

.on('cycle', function (event, bench) {
  console.log(String(bench));
})
.on('complete', function () {
  console.log('Fastest is ' + this.filter('fastest').pluck('name'));
})
.run({async: true, delay: 5});



benchmark/parse_getrow_result.js
/*!
 * ots - benchmark/parse_getrow_result.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var fs = require('fs');
var path = require('path');
var Schema = require('protobuf-ali').Schema;
var Benchmark = require('benchmark');

var root = path.dirname(__dirname);
var schema = new Schema(fs.readFileSync(path.join(root, 'ots_protocol.desc')));
var GetRowResponse = schema['com.aliyun.cloudservice.ots.GetRowResponse'];
var GetRowRequest = schema['com.aliyun.cloudservice.ots.GetRowRequest'];

var requestBuffer = fs.readFileSync(path.join(root, 'test', 'fixtures', 'getRow_request.pb'));
var responseBuffer = fs.readFileSync(path.join(root, 'test', 'fixtures', 'getRow_result.pb'));
var responseEmptyBuffer = fs.readFileSync(path.join(root, 'test', 'fixtures', 'getRow_result_empty.pb'));

console.log('GetRowRequest %j', GetRowRequest.parse(requestBuffer));
console.log('GetRowResponse %j', GetRowResponse.parse(responseBuffer));
console.log('GetRowResponse empty %j', GetRowResponse.parse(responseEmptyBuffer));

var suite = new Benchmark.Suite();

suite
.add('GetRowRequest.parse(requestBuffer)', function () {
  GetRowRequest.parse(requestBuffer);
})
.add('GetRowResponse.parse(responseBuffer)', function () {
  GetRowResponse.parse(responseBuffer);
})
.add('GetRowResponse.parse(responseEmptyBuffer)', function () {
  GetRowResponse.parse(responseEmptyBuffer);
})

.on('cycle', function (event) {
  console.log(String(event.target));
})
.on('complete', function () {
  console.log('Fastest is ' + this.filter('fastest').pluck('name'));
})
.run();


benchmark/testMaxKeepAliveRequests.js
/*!
 * ots - test max keep alive requests
 * http://httpd.apache.org/docs/current/mod/core.html#maxkeepaliverequests
 * 
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var should = require('should');
var Agent = require('agentkeepalive');
var ots = require('../');
var config = require('../test/config.json');

var agent = new Agent();
var client = ots.createClient({
  accessID: config.accessID,
  accessKey: config.accessKey,
  agent: agent
});
var name = 'service.ots.aliyun.com:80';

function start() {
  // client.putData('testbenchmark', {Name: 'name', Value: 'mk2'},
  //   [
  //     {Name: 'lastname', Value: 'feng\' aerdeng' },
  //     {Name: 'nickname', Value: '  苏千\n ' }
  //   ], function (err, result) {
  //   console.log(arguments);
  // });

  var count = 0;

  function request(callback) {
    count++;
    client.getRow('testbenchmark', { Name: 'name', Value: 'mk2' }, function (err, row) {
      if (count % 10 !== 0 && count % 10 !== 1) {
        return callback();
      }
      console.log(count, row);
      console.log('%d sockets, %d socket created, %d requested',
        agent.sockets[name].length,
        agent.createSocketCount,
        agent.requestFinishedCount);
      callback();
    });
  }

  function next() {
    request(next);
  }

  next();
}

client.createTable({
  TableName: 'testbenchmark',
  PrimaryKey: [
    {'Name': 'name', 'Type': 'STRING'},
  ]
}, function (err, result) {
  console.log(arguments);
  start();
});


benchmark/bench.js
/*!
 * ots - benchmark/bench.js
 * Copyright(c) 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

var utility = require('utility');
var ab = require('ab');
var Agent = require('agentkeepalive');
var ots = require('../');
var config = require('../test/config');

var method = process.argv[2];
if (!method) {
  console.log('$ node bench.js [get|put] [concurrency=5] [requests=10000] [keepalive=1]');
  process.exit(0);
}

var concurrency = parseInt(process.argv[3], 10) || 5;
var requests = parseInt(process.argv[4], 10) || 10000;
var options = {
  concurrency: concurrency,
  requests: requests
};

var keepalive = process.argv[5] === '1';

console.log('%s() benchmark with concurrency: %d, requests: %d, keepalive: %s\n', method, concurrency, requests, keepalive);

var client = ots.createClient({
  accessID: config.accessID,
  accessKey: config.accessKey,
  APIHost: config.APIHost,
  agent: keepalive ? new Agent() : null
});

var key = 'ots-bench-key';

function get() {
  var index = 0;
  ab.run(function (callback) {
    var i = index++;
    var uid = utility.md5(key + i);
    client.getRow('tcif', [{Name: 'md5', Value: uid}, {Name: 'user_id', Value: i}], 
    function (err, row) {
      callback(err, row && row.uid === uid && row.firstname === ('yuan-' + i));
    });
  }, options).on('end', function () {
    client.close();
  }).on('error', function (err) {
    console.error(err);
  });
}

function put() {
  var index = 0;
  ab.run(function (callback) {
    var i = index++;
    var uid = utility.md5(key + i);
    client.putRow('tcif', 
      [ 
        { Name: 'md5', Value: uid }, 
        { Name: 'user_id', Value: i },
      ],
      [
        { Name: 'lastname', Value: 'feng\' mk2' },
        { Name: 'nickname', Value: '  苏千\n ' },
        { Name: 'age', Value: 28 }, // int64
        { Name: 'json', Value: '{ "foo": "bar" }' },
        { Name: 'price', Value: 110.5 },
        { Name: 'enable', Value: true },
        { Name: 'man', Value: true },
        { Name: 'status', Value: null },
        { Name: 'female', Value: false },
        { Name: 'createtime', Value: new Date().toJSON() },
        { Name: 'zero', Value: 0 }
      ], 
      function (err) {
        callback(err, err ? false : true);
      }
    );
  }, options).on('end', function () {
    client.close();
  }).on('error', function (err) {
    console.error(err);
  });
}
var set = put;

client.getTableMeta('tcif', function (err, result) {
  console.log(err, result);
  client.listTable(function (err, tables) {
    if (err) {
      throw err;
    }
    eval(method + '();');
  });
});
